extensions/yarn/main.ts

```diff
@@ -85,9 +92,57 @@ if (Deno.args.length >= 1
         || Deno.args[0] === 'dedupe'
    ) {
     await checkDryRun(Deno.args[0], Deno.args.slice(1));
-}
 
-// Run the command with side effects.
-console.log(`[${green("phylum")}] Applying changes…`);
-let status = await Deno.run({ cmd: ['yarn', ...Deno.args] }).status();
-Deno.exit(status.code);
+    console.log(`[${green("phylum")}] Downloading packages to cache…`);
+
+    // Download packages to cache without sandbox.
+    let status = await Deno.run({ cmd: ['yarn', ...Deno.args, '--mode=skip-build']}).status();
+
+    // Ensure download worked. Failure is still "safe" for the user.
+    if (!status.success) {
+        console.error(`[${red("phylum")}] Downloading packages to cache failed.\n`);
+        Deno.exit(status.code);
+    } else {
+        console.log(`[${green("phylum")}] Cache updated successfully.\n`);
+    }
+
+    console.log(`[${green("phylum")}] Building packages inside sandbox…`);
+
+    // Run build inside a sandbox.
+    const output = PhylumApi.runSandboxed({
+        cmd: 'yarn',
+        args: ['install', '--immutable', '--immutable-cache'],
+        exceptions: {
+            run: ['yarn', 'node'],
+            read: ['./', '~/.cache/node/corepack'],
+            write: ['./.yarn/install-state.gz'],
+            net: false,
+        },
+    });
+
+    // Failure here could indicate vulnerabilities; report to the user.
+    if (!output.success) {
+        console.log(`[${red("phylum")}] Sandboxed build failed.`);
+        console.log(`[${red("phylum")}]`);
+
+        // Check for SIGSYS, since this confirms sandbox violation.
+        if (output.signal === 31) {
```

Comment:
Checking for a specific signal at this high level feels a bit weird. Maybe it would make more sense to have a boolean field that `runSandbox` sets when a sanbox violation is detected (something like `output.violation`)?

Otherwise we may have problems when we abstract this to different platforms, since (for example) macos may use a different signal.

Reply:
I don't know what macOS does at all, that's kinda why I'm exposing just the process level API here.

As far as I know, there's no way to reliably detect a sandbox violation so I don't think it would be a good idea to provide an API to read this. Since that API is going to be unreliable and thus misleading.

Reading the `SIGSYS` signal is a heuristic at best.

Reply:
It may be just a heuristic, but it's important enough that we are printing different messages for this extension (and likely many others). I'm just worried that these heuristic checks might get out of hand. But I guess we can make that determination after we figure out macos.

Reply:
My concern is just that when we put an `output.violation` field, people are going to assume there's been a violation when it's `true` and there hasn't been one when it's `false`.

But we can't say that for sure in either direction.

That's why we aren't actually doing anything different. We're just adapting our logging to provide the best possible feedback.

Reply:
I share the same concern. But it seems like a naming problem more than anything else...

We could name it `output.violation_likely`... Or we could make it something other than a bool. Like a string that gets set to either `"likely"` or `"¯\_(ツ)_/¯"` depending on the heuristic.

Reply:
I like the shrug. :smile: